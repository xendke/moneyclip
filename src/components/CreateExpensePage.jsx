import React from 'react';

const CreateExpensePage = () => (
	<div>
		<header>
			<h3>Create</h3>
		</header>
		<p>Expenses... Expenses...</p>
	</div>
);

export default CreateExpensePage;
