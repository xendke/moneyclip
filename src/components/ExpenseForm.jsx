import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			note: props.expense ? props.expense.note : '',
			date: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: ''
		}
	}
	onDescriptionChange = (e) => {
		const value = e.target.value;
		this.setState(() => ({ description: value }));
	}
	onAmountChange = (e) => {
		const value = e.target.value;
		if (!value || value.match(/^\d{1,}(\.\d{0,2})?$/)) { // update only if in format 0000.00
			this.setState(() => ({ amount: value }));
		}
	}
	onNoteChange = (e) => {
		const value = e.target.value;
		this.setState(() => ({ note: value }));
	}
	onDateChange = (moment) => {
		if (!moment) return
		this.setState(() => ({ date: moment }));
	}
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));
	}
	onSubmit = (e) => {
		e.preventDefault();
		if (!this.state.amount || !this.state.description) {
			this.setState(() => ({ error: 'amount and description cannot be empty' }));
		} else {
			this.setState(() => ({ error: '' }));
			this.props.onSubmit({
				id: this.props.expense.id,
				description: this.state.description,
				amount: parseFloat(this.state.amount) * 100,
				note: this.state.note,
				createdAt: this.state.date.valueOf()
			});
		}
	}
	render() {
		return (
			<div>
				{this.state.error ? <p>{this.state.error}</p> : null}
				<form onSubmit={this.onSubmit}>
					<input
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type="text"
						placeholder="Amount"
						autoFocus
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<textarea
						placeholder="More notes (Optional)"
						value={this.state.note}
						onChange={this.onNoteChange}
					/>
					<SingleDatePicker
						date={this.state.date}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						id="ExpenseForm__Calendar"
						numberOfMonths={1}
						isOutsideRange={(day) => (false)}
					/>
					<button>Add Expense</button>
					<button onClick={this.props.onRemoveClick}>remove</button>
				</form>
			</div>
		);
	}
}

export default ExpenseForm;