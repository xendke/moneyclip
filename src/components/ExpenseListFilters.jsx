import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../redux/actions';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null // because there are two date pickers, value is not just true/false
	}
	onFocusChange = (focus) => {
		this.setState(() => ({ calendarFocused: focus }));
	}
	render() {
		const { filters, onTextChange, onSortChange, onDatesChange } = this.props;
		return (
			<div>
				<input type="text" value={filters.text} onChange={onTextChange} />
				Sort By:
				<select value={filters.sortBy} onChange={onSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					onDatesChange={onDatesChange}
					startDate={filters.startDate ? filters.startDate : undefined}
					endDate={filters.endDate ? filters.endDate : undefined}
					startDateId="ExpenseListFilters__DateRangePicker__StartDate"
					endDateId="ExpenseListFilters__DateRangePicker__EndDate"
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					isOutsideRange={() => (false)}
					numberOfMonths={1}
					showClearDates={true}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => {
	return {
		filters: state.filters
	}
}
const mapDispatchToProps = (dispatch) => {
	return {
		onTextChange: (e) => {
			dispatch(setTextFilter(e.target.value));
		},
		onSortChange: (e) => {
			if (e.target.value === 'date') {
				dispatch(sortByDate());
			} else {
				dispatch(sortByAmount());
			}
		},
		onDatesChange: ({ startDate, endDate }) => {
			dispatch(setStartDate({ startDate }));
			dispatch(setEndDate({ endDate }));
		}
	}
}

const ConnectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);


export default ConnectedExpenseListFilters;