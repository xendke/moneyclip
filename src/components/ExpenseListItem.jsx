import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../redux/actions';

const ExpenseListItem = (props) => (
	<div>
		<h4>{props.description}</h4>
		<p>{props.amount} - {props.createdAt}</p>
		<button onClick={props.onRemoveClick}>remove</button>
	</div>
);

const mapDispatchToProps = (dispatch, ownProps) => {
	return {
		onRemoveClick: () => { dispatch(removeExpense({ id: ownProps.id })) }
	}
}

const ConnectedExpenseListItem = connect(undefined, mapDispatchToProps)(ExpenseListItem);

export default ConnectedExpenseListItem;