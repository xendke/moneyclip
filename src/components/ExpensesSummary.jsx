import React from 'react';
import { connect } from 'react-redux';
import { getVisibleExpenses, getTotalFromExpenses } from '../redux/selectors';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
	const expenseWord = expenseCount > 1 ? 'expenses' : 'expense';
	const total = numeral(expensesTotal / 100).format('$0,0.00');

	return (
		<p>
			{!expenseCount || expenseCount <= 0 ?
				<span>Viewing no expenses. Try clearing the filters.</span>
				:
				<span >Viewing {expenseCount} {expenseWord} for a total of {total}.</span>
			}
		</p>
	);
};

const mapStateToProps = (state) => {
	const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
	return {
		expenseCount: visibleExpenses.length,
		expensesTotal: getTotalFromExpenses(visibleExpenses)
	};
}

const ConnectedExpensesSummary = connect(mapStateToProps)(ExpensesSummary);

export default ConnectedExpensesSummary;