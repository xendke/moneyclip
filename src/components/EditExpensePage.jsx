import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../redux/actions';

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.editExpense(expense);
		this.props.history.push('/');
	}
	onRemoveClick = (e) => {
		e.preventDefault();
		this.props.removeExpense(this.props.match.params.id);
		this.props.history.push('/');
	}
	render() {
		return (
			<div>
				<header>
					<h3>Edit</h3>
				</header>
				<ExpenseForm
					expense={this.props.expense}
					onSubmit={this.onSubmit}
					onRemoveClick={this.onRemoveClick}
				/>
			</div>
		)
	}
}

const mapStateToProps = (state, ownProps) => {
	return {
		expense: state.expenses.find((expense) => {
			return expense.id === ownProps.match.params.id;
		})
	}
}
const mapDispatchToProps = (dispatch) => {
	return {
		editExpense: (expense) => {
			dispatch(editExpense(expense));
		},
		removeExpense: (id) => {
			dispatch(removeExpense({ id }));
		}
	}
}
const ConnectedEditExpensePage = connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
export default ConnectedEditExpensePage;
