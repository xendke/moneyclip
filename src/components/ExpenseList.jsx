import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import visibleExpenses from '../redux/selectors';

export const ExpenseList = (props) => (
	<div>
		{
			props.expenses.length === 0 ? (
				<p>No expenses yet!</p>
			) : (
					props.expenses.map((expense) => (
						<ExpenseListItem key={expense.id} {...expense} />
					))
				)
		}
	</div>
);

const mapStateToProps = (state) => {
	return {
		expenses: visibleExpenses(state.expenses, state.filters)
	};
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);
export default ConnectedExpenseList;