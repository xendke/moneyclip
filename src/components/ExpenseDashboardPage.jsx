import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
	<div>
		{/* <header>
			<h3>Dashboard</h3>
		</header> */}
		<ExpenseListFilters />
		<ExpenseList />
	</div>
);

export default ExpenseDashboardPage;
