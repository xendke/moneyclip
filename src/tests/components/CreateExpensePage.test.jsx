import React from 'react';
import { shallow } from 'enzyme';
import { CreateExpensePage } from '../../components/CreateExpensePage';
import expenses from '../fixtures/expenses';

let onSubmitSpy, mockHistory, wrapper;
beforeEach(() => {
	onSubmitSpy = jest.fn();
	mockHistory = { push: jest.fn() };
	wrapper = shallow(<CreateExpensePage onFormSubmit={onSubmitSpy} history={mockHistory} />);
});

it('renders CreateExpensePage correctly', () => {
	expect(wrapper).toMatchSnapshot();
});

it('handles onSubmit', () => {
	wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
	expect(onSubmitSpy).toHaveBeenCalledWith(expenses[0]);
	expect(mockHistory.push).toHaveBeenCalledWith('/');
});
