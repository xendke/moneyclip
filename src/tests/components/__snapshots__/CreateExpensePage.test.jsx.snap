// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CreateExpensePage correctly 1`] = `
<div>
  <header>
    <h3>
      Create
    </h3>
  </header>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
